### ISSUE TITLE
Need Externals Libs and only work with vs2015 or lower

### ISSUE THREAD

**Issue by GordonEldest** [cr333/usac-cmake#2]
This project is pretty much useless
A- there is no sample and not config file so it is of no interest to even have a RunTest as there is no test

B- The link to otiginal code is broken
http://www.cs.unc.edu/~rraguram/usac/
So you will not have access to any additional info

To compile it
You need:

- VS2015 (anything higher give issue due to  error LNK2001: unresolved external symbol __iob_func failed by MinGW to be up-todate
You will need either to learn how to compile Lapack etc ... 
or use prebuild

- Also you need additional projects 
https://github.com/hyperrealm/libconfig

- And Lapack and Blas
http://icl.cs.utk.edu/lapack-for-windows/clapack/index.html

- And 
https://www.netlib.org/clapack/LIB_WINDOWS/prebuilt_libraries_windows.html

But as explained it only work until vs2015 due to MinGw and prebuilt not build adequately.
Need to resolve linker error (google it) you may need to install MinGW and recompile Lapack

In short:
Be ready to spend a day to link that. 

Feasible but 
Better go on OpenCV and use USAC from something that doesn't use Lapack (Lapack guys are too hostile to Microsoft)

That should be good to put theses info im md

**Comment by cr333** [comment]
Thank you for taking the time to write up the issues you encountered. I have not used this code recently, so I was not aware of this.

As stated in the license file, this code is "provided 'as is' without express or implied warranty." The dependencies you mention are the dependencies of the original USAC code, not this port to CMake.

It looks like USAC was added to OpenCV as recently as version 4.5 (October 2020). I would probably use this in future.

**Comment by GordonEldest** [comment]
Thx

I linked it and make it going to main with VS 2019 (see below how)

Frustrating because it looks like a good code, healthier than the OCV implementation that are made for OCV objects (Generally Salient points) hence  this USAC implementation seems more general purpose, or at least easier to derive your own class.
The Algo is still up-to-date and USAC is new on OCV so ... this implementation worth a try. 

You need to install clapack instead of lapack, and blas,
Take CLAPACK-3.1.1-VisualStudio from https://www.netlib.org/clapack/
Need some work but not so much.

What is lacking is a baseline test set . The read.me is not enough and there is no way to be sure it works at it best and there is no explanation on the API.

the Main need 2 files
RunSingleTest <estimation problem> <config file>
<estimation problem>: 0 (fundamental matrix), 1 (homography)"
<config file>: full path to configuration file"

The paper about it http://people.inf.ethz.ch/pomarc/pubs/RaguramPAMI13.pdf
Very clear that USAC is better to resolve critical condition. Better than PROSAC.

If somebody have an example of a "config" file and a reference "test set" that would be greatly appreciated and beneficial to a good code

Note:
Link with MinGW is pretty tedious.
I will not debate who is wrong but lapack peoples seems without intent to maintain an uptodate version of the compiled lib (they only use MinGW or GCC) compatible with new MS convention for __iob_fun. 
Seems part of the anti-Microsoft front that start to be really boring.
Hence my personal choice to reject code that are non-MS compatible, generally they prefigure issues with Unix too.

**Comment by cr333** [comment]
Thanks for your perseverance. I think I compiled CLAPACK with MSVC to avoid linking issues.

I have looked for the original data, and uploaded what I could find here:
https://github.com/cr333/usac-cmake/releases/download/USAC-1.0/USAC-1.0-code-and-data.zip

This also contains copies of CLAPACK 3.2.1 and libconfig 1.4.5, which I believe were part of the original USAC 1.0 release.

Feel free to contribute to this repo if you find ways to improve the code and/or documentation.

Hope this helps.

**Comment by GordonEldest** [comment]
Thank you
It helps tremendously!
I guess it's all on board now. After a quick glance at the config file, it's clear that I wouldn't have been able to guess it without a long session of reverse-engineering, debugging the code intent.
So you deliver here a big time saver!

As my project under VS2019 is already compiled and linked, I will test it before checking the integrity of the Cmake / compile / Link on the supplied assembly. But at first glance, it seems like it's all there including a more recent version of CLAPACK

As I explained, this USAC code seems more appropriate to derive usage.
I consider it for geothermal data monitoring and analysis, and implementing the OCV USAC requires more than a quirk to use it adapt OCV salientpoint data objects.
